src/util/heap/blockpageresource.rs

```diff
@@ -0,0 +1,417 @@
+        (*self.data.get())[i] = block
+    }
+
+    /// Non-atomically push an element.
+    ///
+    /// It's unsafe unless the array is accessed by only one thread (i.e. used as a thread-local array).
+    #[inline(always)]
+    unsafe fn push_relaxed(&self, block: Block) -> Result<(), Block> {
+        let i = self.cursor.load(Ordering::Relaxed);
+        if i < self.capacity {
+            self.set_entry(i, block);
+            self.cursor.store(i + 1, Ordering::Relaxed);
+            Ok(())
+        } else {
+            Err(block)
+        }
+    }
+
+    /// Atomically pop an element from the array.
+    #[inline(always)]
+    fn pop(&self) -> Option<Block> {
+        let i = self
+            .cursor
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |i| {
+                if i > 0 {
+                    Some(i - 1)
+                } else {
+                    None
+                }
+            });
+        if let Ok(i) = i {
+            Some(self.get_entry(i - 1))
+        } else {
+            None
+        }
+    }
+
+    /// Get array size
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.cursor.load(Ordering::SeqCst)
+    }
+
+    /// Test if the array is empty
+    #[inline(always)]
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Iterate all elements in the array
+    #[inline]
+    fn iterate_blocks(&self, f: &mut impl FnMut(Block)) {
+        let len = self.len();
+        for i in 0..len {
+            f(self.get_entry(i))
+        }
+    }
+
+    /// Replace the array with a new array.
+    ///
+    /// Return the old array
+    #[inline(always)]
+    fn replace(&self, new_array: Self) -> Self {
+        debug_assert_eq!(self.capacity, new_array.capacity);
+        // Swap cursor
+        let temp = self.cursor.load(Ordering::Relaxed);
+        self.cursor
+            .store(new_array.cursor.load(Ordering::Relaxed), Ordering::Relaxed);
+        new_array.cursor.store(temp, Ordering::Relaxed);
+        // Swap data
+        unsafe {
+            std::mem::swap(&mut *self.data.get(), &mut *new_array.data.get());
+        }
+        // Return old array
+        new_array
+    }
+}
+
+/// A block queue which contains a global pool and a set of thread-local queues.
+///
+/// Mutator or collector threads always allocate blocks by poping from the global pool。
+///
+/// Collector threads free blocks to their thread-local queues, and then flush to the global pools before GC ends.
+pub struct BlockQueue<Block> {
+    /// First global BlockArray for fast allocation
+    head_global_freed_blocks: RwLock<Option<BlockArray<Block>>>,
+    /// A list of BlockArray that is flushed to the global pool
+    global_freed_blocks: RwLock<Vec<BlockArray<Block>>>,
+    /// Thread-local block queues
+    worker_local_freed_blocks: Vec<BlockArray<Block>>,
+    /// Total number of blocks in the whole BlockQueue
+    count: AtomicUsize,
+}
```

Comment:
I suggest renaming this to `BlockPool`, and optionally rename `BlockArray` to `BlockQueue`. Having both `BlockArray` and `BlockQueue` is a bit confusing.

Reply:
Renamed as follows:
* `BlockQueue` => `BlockPool`
* `BlockArray` => `BlockQueue`
* `BlockList` => `ReusableBlockPool`